File: apps/passenger-android/src/main/kotlin/lk/mageride/passenger/settings/EditProfileScreen.kt
package lk.mageride.passenger.settings

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Delete
import androidx.compose.material.icons.filled.Edit
import androidx.compose.material.icons.filled.Person
import androidx.compose.material.icons.filled.PhotoCamera
import androidx.compose.material3.AlertDialog
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.OutlinedButton
import androidx.compose.material3.Switch
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.text.input.ImeAction
import androidx.compose.ui.text.input.KeyboardType
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import lk.mageride.passenger.R
import lk.mageride.passenger.onboarding.PhoneNumber
import lk.mageride.passenger.ui.component.InlineError
import lk.mageride.passenger.ui.component.LabelledTextField
import lk.mageride.passenger.ui.component.MageRideTextLink
import lk.mageride.passenger.ui.component.PhoneNumberField
import lk.mageride.passenger.ui.component.SectionLabel
import lk.mageride.passenger.ui.theme.ControlTokens
import lk.mageride.passenger.ui.theme.MageRideTheme
import lk.mageride.shared.data.models.iam.EmergencyContact

/**
 * SCR-PA-027b — "Edit profile".
 *
 * The wireframe: `‹ Edit profile … Save`, the avatar with its 📷 badge and *"Take photo or upload"*,
 * **Full name**, the **Notifications & offers** switch, and the SOS contact list over
 * `＋ Add SOS contact`.
 *
 * **No language row** (AL-26) — the wireframe says so, and the view model's save has no field for
 * one.
 *
 * **The photo control is drawn and inert, and that is a missing route rather than a missing
 * screen.** `UpdateProfileRequest.photoUrl` is a *URL*, and nothing on the app-facing surface mints
 * one for a passenger: `POST /v1/support/screenshots`, `POST /v1/mode-b/payments/{id}/transfer-slip`
 * and the driver's document uploads are the whole upload set, and none of them is an avatar. C077
 * hit the same wall on SCR-PA-004 and made the same call; landing it needs an upload route first.
 * Disabled rather than silently ignored, because a control that looks live and does nothing is
 * worse than one that says it is not ready. Recorded in the C083 handoff.
 */
@Composable
@Suppress("LongMethod") // The screen is the wireframe's layout tree, top to bottom.
internal fun EditProfileScreen(onBack: () -> Unit, model: EditProfileViewModel) {
    val state by model.state.collectAsStateWithLifecycle()

    LaunchedEffect(state.saved) {
        if (state.saved) onBack()
    }

    Column(modifier = Modifier.fillMaxSize()) {
        SettingsTopBar(
            title = stringResource(R.string.edit_profile_title),
            onBack = onBack,
            actions = {
                MageRideTextLink(
                    label = stringResource(R.string.action_save),
                    onClick = model::save,
                    enabled = state.canSave,
                )
            },
        )

        Column(
            modifier = Modifier
                .fillMaxSize()
                .verticalScroll(rememberScrollState())
                .padding(horizontal = MageRideTheme.spacing.md)
                .padding(bottom = MageRideTheme.spacing.md),
            verticalArrangement = Arrangement.spacedBy(MageRideTheme.spacing.sm),
        ) {
            state.error?.let { InlineError(stringResource(it)) }

            AvatarEditor()

            LabelledTextField(
                label = stringResource(R.string.edit_profile_name),
                value = state.name,
                onValueChange = model::onNameChanged,
                enabled = state.loaded && !state.saving,
                imeAction = ImeAction.Done,
            )

            Row(modifier = Modifier.fillMaxWidth(), verticalAlignment = Alignment.CenterVertically) {
                Text(
                    text = stringResource(R.string.edit_profile_notifications),
                    modifier = Modifier.weight(1f),
                    style = MaterialTheme.typography.bodyLarge,
                    color = MaterialTheme.colorScheme.onSurface,
                )
                Switch(
                    checked = state.notificationsEnabled,
                    onCheckedChange = model::onNotificationsChanged,
                    enabled = state.loaded && !state.saving,
                )
            }

            SectionLabel(
                text = stringResource(R.string.edit_profile_sos_contacts),
                modifier = Modifier.padding(top = MageRideTheme.spacing.xxs),
            )

            if (state.contacts.isEmpty()) {
                // What an empty list costs, said here rather than at the moment of an alarm:
                // `POST /v1/sos` answers `400 no-emergency-contact` with none on file.
                Text(
                    text = stringResource(R.string.edit_profile_sos_empty),
                    style = MaterialTheme.typography.bodySmall,
                    color = MaterialTheme.colorScheme.onSurfaceVariant,
                )
            }

            state.contacts.forEach { contact ->
                ContactRow(
                    contact = contact,
                    busy = state.removing == contact.contactId,
                    onEdit = { model.editContact(contact) },
                    onRemove = { model.removeContact(contact) },
                )
            }

            OutlinedButton(
                onClick = model::addContact,
                modifier = Modifier.fillMaxWidth(),
                enabled = state.loaded,
            ) {
                Text(stringResource(R.string.edit_profile_add_sos))
            }
        }
    }

    state.contactDraft?.let { draft ->
        ContactDialog(
            draft = draft,
            onNameChanged = model::onContactNameChanged,
            onPhoneChanged = model::onContactPhoneChanged,
            onSave = model::saveContact,
            onDismiss = model::dismissContact,
        )
    }
}

/**
 * The wireframe's `avatar lg` with its 📷 badge and *"Take photo or upload"* textlink.
 *
 * Both are inert — see the screen's KDoc. The photo itself is not rendered either: the profile
 * carries a URL and this app ships no image loader (C082 argued the same point about the owner's
 * LankaQR image and decoded bytes by hand rather than adding one).
 */
@Composable
private fun AvatarEditor() {
    Column(
        modifier = Modifier.fillMaxWidth(),
        horizontalAlignment = Alignment.CenterHorizontally,
        verticalArrangement = Arrangement.spacedBy(MageRideTheme.spacing.xs),
    ) {
        Box(contentAlignment = Alignment.BottomEnd) {
            Box(
                modifier = Modifier
                    .size(ControlTokens.AvatarLarge)
                    .background(MaterialTheme.colorScheme.surfaceVariant, CircleShape),
                contentAlignment = Alignment.Center,
            ) {
                Icon(
                    imageVector = Icons.Filled.Person,
                    contentDescription = null,
                    modifier = Modifier.size(ControlTokens.IllustrationIcon),
                    tint = MaterialTheme.colorScheme.onSurfaceVariant,
                )
            }
            Box(
                modifier = Modifier
                    .size(ControlTokens.AvatarBadge)
                    .background(MaterialTheme.colorScheme.surface, CircleShape),
                contentAlignment = Alignment.Center,
            ) {
                Icon(
                    imageVector = Icons.Filled.PhotoCamera,
                    contentDescription = null,
                    modifier = Modifier.size(ControlTokens.ChipIcon),
                    tint = MaterialTheme.colorScheme.outlineVariant,
                )
            }
        }

        MageRideTextLink(
            label = stringResource(R.string.edit_profile_photo),
            onClick = {},
            enabled = false,
        )
    }
}

/** One SOS contact — the wireframe's `👤 Amma · +94 77 000 1111 ✎`, plus the delete US-12.1 needs. */
@Composable
private fun ContactRow(contact: EmergencyContact, busy: Boolean, onEdit: () -> Unit, onRemove: () -> Unit) {
    SettingsRow(
        icon = Icons.Filled.Person,
        label = contact.name,
        supporting = contact.phone,
        onClick = onEdit,
        trailing = {
            IconButton(onClick = onEdit, enabled = !busy) {
                Icon(
                    imageVector = Icons.Filled.Edit,
                    contentDescription = stringResource(R.string.edit_profile_edit_sos),
                    modifier = Modifier.size(ControlTokens.RowIcon),
                    tint = MaterialTheme.colorScheme.onSurfaceVariant,
                )
            }
            IconButton(onClick = onRemove, enabled = !busy) {
                Icon(
                    imageVector = Icons.Filled.Delete,
                    contentDescription = stringResource(R.string.edit_profile_remove_sos),
                    modifier = Modifier.size(ControlTokens.RowIcon),
                    tint = MaterialTheme.colorScheme.error,
                )
            }
        },
    )
}

/**
 * The add/edit contact dialog.
 *
 * The number is entered through the same `+94`-prefixed field SCR-PA-003 uses, because it is the
 * same question: Sri Lanka is the only country this platform operates in, so the prefix is a label
 * on the field rather than a country picker, and both `0771234567` and `771234567` have to work.
 */
@Composable
private fun ContactDialog(
    draft: SosContactDraft,
    onNameChanged: (String) -> Unit,
    onPhoneChanged: (String) -> Unit,
    onSave: () -> Unit,
    onDismiss: () -> Unit,
) {
    AlertDialog(
        onDismissRequest = onDismiss,
        title = {
            Text(
                stringResource(
                    if (draft.contactId == null) R.string.edit_profile_add_sos else R.string.edit_profile_edit_sos,
                ),
            )
        },
        text = {
            Column(verticalArrangement = Arrangement.spacedBy(MageRideTheme.spacing.xs)) {
                LabelledTextField(
                    label = stringResource(R.string.edit_profile_sos_name),
                    value = draft.name,
                    onValueChange = onNameChanged,
                    enabled = !draft.saving,
                    keyboardType = KeyboardType.Text,
                )
                SectionLabel(stringResource(R.string.edit_profile_sos_phone))
                PhoneNumberField(
                    value = draft.phone,
                    onValueChange = onPhoneChanged,
                    countryCode = PhoneNumber.COUNTRY_CODE,
                    placeholder = PhoneNumber.PLACEHOLDER,
                    enabled = !draft.saving,
                )
            }
        },
        confirmButton = {
            TextButton(onClick = onSave, enabled = draft.canSave) {
                Text(stringResource(R.string.action_save))
            }
        },
        dismissButton = { TextButton(onClick = onDismiss) { Text(stringResource(R.string.action_cancel)) } },
    )
}

File: apps/passenger-ios/PassengerApp/Settings/EditProfileScreen.swift
import MageRideShared
import SwiftUI

/// SCR-PI-027b — *"Edit profile"*.
///
/// The cell: `‹ Settings · Edit profile · **Save**`, the `avatar lg` with its 📷 badge and *"Take
/// photo or upload"*, a `glist` holding **Full name**, a second holding the **Notifications &
/// offers** switch, the `SOS / emergency contacts` label, and a third `glist` of contact rows over a
/// centred `＋ Add SOS contact`.
///
/// **No language row** (AL-26) — the cell says so, and the model's save has no field for one.
///
/// **The photo control is drawn and inert, and that is a missing route rather than a missing
/// screen.** `UpdateProfileRequest.photoUrl` is a *URL*, and nothing on the app-facing surface mints
/// one for a passenger: `POST /v1/support/screenshots`, the Mode B transfer slip and the driver's
/// document uploads are the whole upload set, and none of them is an avatar. C095 hit the same wall
/// on SCR-PI-004 and made the same call; landing it needs an upload route first. Disabled rather
/// than silently ignored, because a control that looks live and does nothing is worse than one that
/// says it is not ready. The cell's `PhotosPicker` note is therefore unimplementable today and is
/// recorded in the C101 handoff.
///
/// **Save is in the navigation bar**, which is what the cell's `.act` slot draws and what
/// SCR-PI-004 already does on this platform.
@MainActor
struct EditProfileScreen: View {

    @StateObject private var model: EditProfileModel

    /// Popped on a successful save — the wireframe's `‹ Settings` is where a saved profile lands.
    let onSaved: () -> Void

    init(
        profiles: PassengerProfileRepository,
        contacts: SosContacts,
        identity: PassengerIdentity,
        keys: IdempotencyKeys,
        onSaved: @escaping () -> Void
    ) {
        _model = StateObject(
            wrappedValue: EditProfileModel(
                profiles: profiles,
                contacts: contacts,
                identity: identity,
                keys: keys
            )
        )
        self.onSaved = onSaved
    }

    var body: some View {
        ScrollView {
            VStack(alignment: .leading, spacing: MageRideSpacing.sm) {
                if let errorKey = model.state.errorKey {
                    FormErrorText(messageKey: errorKey)
                }

                avatar

                GroupedList {
                    GroupedRow(titleKey: "edit_profile_name", showsSeparator: false) {
                        TextField("edit_profile_name".localised, text: nameBinding)
                            .mageFont(.body)
                            .foregroundStyle(MageRideColor.onSurface)
                            .multilineTextAlignment(.trailing)
                            .textContentType(.givenName)
                            .textInputAutocapitalization(.words)
                            .disabled(!model.state.isLoaded || model.state.isSaving)
                    }
                }

                GroupedList {
                    GroupedRow(titleKey: "edit_profile_notifications", showsSeparator: false) {
                        Toggle("", isOn: notificationsBinding)
                            .labelsHidden()
                            .disabled(!model.state.isLoaded || model.state.isSaving)
                    }
                }

                SectionLabel(key: "edit_profile_sos_contacts")
                    .padding(.top, MageRideSpacing.xxs)

                if model.state.contacts.isEmpty {
                    // What an empty list costs, said here rather than at the moment of an alarm:
                    // `POST /v1/sos` answers `400 no-emergency-contact` with none on file.
                    Text(key: "edit_profile_sos_empty")
                        .mageFont(.bodySmall)
                        .foregroundStyle(MageRideColor.onSurfaceVariant)
                }

                GroupedList {
                    ForEach(model.state.contacts, id: \.contactId) { contact in
                        Button {
                            model.editContact(contact)
                        } label: {
                            GroupedValueRow(
                                title: contact.name,
                                subtitle: contact.phone,
                                symbolName: "person.fill",
                                symbolTint: MageRideColor.error
                            ) {
                                // The cell draws `✎` where every other row on this surface draws
                                // `›`, because the row edits rather than navigates.
                                Image(systemName: "pencil")
                                    .font(.system(size: MageRideControl.chipIcon, weight: .semibold))
                                    .foregroundStyle(MageRideColor.onSurfaceVariant)
                                    .accessibilityHidden(true)
                            }
                        }
                        .buttonStyle(.plain)
                        .disabled(model.state.removing == contact.contactId)
                    }

                    CentredActionRow(titleKey: "edit_profile_add_sos", isEnabled: model.state.isLoaded) {
                        model.addContact()
                    }
                }
            }
            .padding(MageRideSpacing.md)
        }
        .background(MageRideColor.surface)
        .navigationTitle(Text(key: "edit_profile_title"))
        .navigationBarTitleDisplayMode(.inline)
        .toolbar {
            ToolbarItem(placement: .navigationBarTrailing) {
                Button {
                    Task { await model.save() }
                } label: {
                    if model.state.isSaving {
                        ProgressView()
                    } else {
                        Text(key: "action_save").mageFont(.subtitle)
                    }
                }
                .disabled(!model.state.canSave)
            }
        }
        .task { await model.load() }
        .onChange(of: model.state.isSaved) { saved in
            if saved { onSaved() }
        }
        .sheet(isPresented: contactBinding) {
            if let draft = model.state.contactDraft {
                SosContactSheet(
                    draft: draft,
                    onNameChanged: model.onContactNameChanged,
                    onPhoneChanged: model.onContactPhoneChanged,
                    onSave: { Task { await model.saveContact() } },
                    onRemove: { Task { await model.removeContact() } }
                )
            }
        }
    }

    // MARK: -

    /// The wireframe's `avatar lg` with its 📷 badge and the *"Take photo or upload"* textlink.
    ///
    /// Both inert — see the screen's note. The photo itself is not rendered either: the profile
    /// carries a URL and this app ships no image loader (C100 argued the same point about the fleet
    /// owner's LankaQR and decoded bytes by hand rather than adding one).
    private var avatar: some View {
        VStack(spacing: MageRideSpacing.xs) {
            ProfileAvatar(badgeSymbol: "camera.fill")
            TextLink(key: "edit_profile_photo", isEnabled: false) { }
        }
        .frame(maxWidth: .infinity)
    }

    private var nameBinding: Binding<String> {
        Binding(get: { model.state.name }, set: model.onNameChanged)
    }

    private var notificationsBinding: Binding<Bool> {
        Binding(get: { model.state.notificationsEnabled }, set: model.onNotificationsChanged)
    }

    private var contactBinding: Binding<Bool> {
        Binding(
            get: { model.state.contactDraft != nil },
            set: { presented in if !presented { model.dismissContact() } }
        )
    }
}

/// The add / edit SOS contact sheet.
///
/// **A `.sheet` where Android has an `AlertDialog`**, for the reason SCR-PI-026a is one: this cluster
/// captures two things and both are short forms with a keyboard, and an alert with two text fields in
/// it is a Material shape rather than an iOS one. The cell draws no frame for this control at all, so
/// the conservative reading is the one its sibling already established.
///
/// The number is entered through the same `+94`-prefixed field SCR-PI-003 uses, because it is the
/// same question: Sri Lanka is the only country this platform operates in, so the prefix is a label
/// on the field rather than a country picker, and both `0771234567` and `771234567` have to work.
private struct SosContactSheet: View {

    let draft: SosContactDraft
    let onNameChanged: (String) -> Void
    let onPhoneChanged: (String) -> Void
    let onSave: () -> Void
    let onRemove: () -> Void

    var body: some View {
        VStack(alignment: .leading, spacing: MageRideSpacing.sm) {
            Text(key: draft.isEditing ? "edit_profile_edit_sos" : "edit_profile_add_sos")
                .mageFont(.title)
                .foregroundStyle(MageRideColor.onSurface)

            LabelledTextField(
                labelKey: "edit_profile_sos_name",
                value: Binding(get: { draft.name }, set: onNameChanged)
            )

            SectionLabel(key: "edit_profile_sos_phone")
            PhoneNumberField(
                value: Binding(get: { draft.phone }, set: onPhoneChanged),
                isEnabled: !draft.isSaving
            )

            Button(action: onSave) {
                Text(key: "action_save")
            }
            .buttonStyle(.mageCta(loading: draft.isSaving))
            .disabled(!draft.canSave)
            .padding(.top, MageRideSpacing.xxs)

            if draft.isEditing {
                Button(action: onRemove) {
                    Text(key: "edit_profile_remove_sos")
                        .mageFont(.bodySmall)
                        .foregroundStyle(MageRideColor.error)
                        .frame(maxWidth: .infinity, minHeight: MageRideControl.minimumTapTarget)
                        .contentShape(Rectangle())
                }
                .buttonStyle(.plain)
                .disabled(draft.isSaving)
            }

            Spacer(minLength: 0)
        }
        .padding(.horizontal, MageRideSpacing.md)
        .padding(.top, MageRideSpacing.md)
        .padding(.bottom, MageRideSpacing.lg)
        .frame(maxWidth: .infinity, alignment: .leading)
        .background(MageRideColor.surface)
        .presentationDetents([.height(MageRideControl.contactSheetHeight), .large])
        .presentationDragIndicator(.visible)
    }
}
